### PR TITLE
Fix framework iterators behavior when iter_setup raises StopIteration

### PR DESCRIPTION
**Pull request by JanuszL**
- when iter_setup raises StopIteration pipeline is no longer feed
  with the data and when it runs out of it StopIteration is raised
  by the framework iterator as well. After reset framework iterator
  has no data ready to return and no more work is scheduled as the
  `__next__` method in it ends prematurely on share_outputs that is empty
- adds an empty method to the pipeline so framework iterator can detect
  if the pipeline is empty and if more work needs to be scheduled ahead in
  reset method

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

#### Why we need this PR?
- bugfix for https://github.com/NVIDIA/DALI/issues/1134

#### What happened in this PR?
 -  adds an empty method to the pipeline so framework iterator can detect
  if the pipeline is empty and if more work needs to be scheduled ahead in
  reset method
 - it changes framework iterator behavior when iter_setup raises StopIteration
 - add examples with ExternalSource and PyTorch and MXNet iterators
 - test was added for this corner case

**JIRA TASK**: DALI-996

### CODE REVIEW

File: dali/test/python/test_FW_iterators_shuffling.py [hunk 1]
```diff
@@ -221,3 +221,70 @@ def test_pytorch_iterator_not_fill_last_batch_pad_last_batch():
...
+            self.feed_input(self.test_data, data)
+        except StopIteration:
+            self.dataset = iter(self.data_source)
+            raise StopIteration
+
+    @property
+    def size(self):
+        return self.data_source.size
+
+def chec_stop_iter(iterator, batch_size, epochs, iter_num):
```
Comment (awolant) [inline dali/test/python/test_FW_iterators_shuffling.py]: Maybe a typo here?

File: dali/test/python/test_FW_iterators_shuffling.py [hunk 2]
```diff
@@ -221,3 +221,70 @@ def test_pytorch_iterator_not_fill_last_batch_pad_last_batch():
...
+              for batch_size in [1, 10 ,100]:
+                  pipe = TestIterPipeline(batch_size, 0, TestIterator(iter_num, batch_size))
+                  loader = fw_iter(pipe, pipe.size)
+                  count = 0
+                  for e in range(epochs):
+                    for i, outputs in enumerate(loader):
+                        count += 1
+                    loader.reset()
+                  assert(count == iter_num * epochs)
+                  yield chec_stop_iter, iter_name, batch_size, epochs, iter_num
```
Comment (awolant) [inline dali/test/python/test_FW_iterators_shuffling.py]: Maybe move the `assert` to `check_stop_iter`? I think it would be more aligned with `nosetest` API?

Comment (JanuszL) [reply dali/test/python/test_FW_iterators_shuffling.py]: Done

File: dali/test/python/test_FW_iterators_shuffling.py (hunk 1 above)
Comment (JanuszL) [reply dali/test/python/test_FW_iterators_shuffling.py]: Done

**Comment by JanuszL** [comment]
> Maybe it would be good to also update the [ExternalSource example](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/external_input.html?highlight=externalsource) ?

Done as well.

**Comment by JanuszL** [comment]
!build

File: dali/python/nvidia/dali/plugin/mxnet.py [hunk 3]
```diff
@@ -286,6 +286,9 @@ def reset(self):
                 self._counter = 0
             for p in self._pipes:
                 p.reset()
+                if p.empty():
+                    with p._check_api_type_scope(types.PieplineAPIType.ITERATOR) as check:
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/python/nvidia/dali/plugin/mxnet.py]: Done

**Comment by JanuszL** [comment]
!build
